### PR TITLE
HDDS-6812. Fix INTERNAL_ERROR message on failed write.

### PR DESCRIPTION
**Pull request by aswinshakil**
## What changes were proposed in this pull request?

When `allocateBlock()` fails, It throws a generic SCMException `Allocated 0 blocks. Requested 1 blocks`. Here the block allocation failed because there weren't enough DataNodes for EC replication. Proposing to throw an exception when `allocateContainer()` fails and propagate back the actual Exception. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6812

## How was this patch tested?

This patch was tested manually with docker and Unit test. Now the error message is,
```
INTERNAL_ERROR No enough datanodes to choose. TotalNode = 3 AvailableNode = 3 RequiredNode = 5 ExcludedNode = 0
```

### CODE REVIEW

**Comment by umamaheswararao** [comment]
Thanks @aswinshakil for working on this. Changes looks good to me. I have a general question: 
try {
      for (int i = 0; i < num; i++) {
        AllocatedBlock block = scm.getScmBlockManager()
            .allocateBlock(size, replicationConfig, owner, excludeList);
        if (block != null) {
          blocks.add(block);
        }
      }
      return blocks;
    } catch (Exception ex) {
    
    Now here it will never return null since you are throwing exception? If yes, probably you may want to file a JIRA to revisit this to remove the if block!=null check and looping?

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestWritableECContainerProvider.java [hunk 1]
```diff
@@ -234,9 +234,11 @@ public Pipeline createPipeline(ReplicationConfig repConf,
     provider = new WritableECContainerProvider(
         conf, pipelineManager, containerManager, pipelineChoosingPolicy);
 
-    ContainerInfo container =
-        provider.getContainer(1, repConfig, OWNER, new ExcludeList());
-    assertNull(container);
+    try {
+      provider.getContainer(1, repConfig, OWNER, new ExcludeList());
```
Comment (umamaheswararao) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestWritableECContainerProvider.java:238]: You have to add Assert.fail here. Lets say if above statement does not throw exception, then case will pass.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestWritableECContainerProvider.java [hunk 2]
```diff
@@ -261,12 +263,18 @@ public Pipeline createPipeline(ReplicationConfig repConf,
...
+    } catch (IOException ex) {
+      GenericTestUtils.assertExceptionContains("Cannot create pipelines", ex);
+    }
+
     for (int i = 0; i < 5; i++) {
-      ContainerInfo nextContainer =
-          provider.getContainer(1, repConfig, OWNER, new ExcludeList());
-      assertEquals(container, nextContainer);
+      try {
+        provider.getContainer(1, repConfig, OWNER, new ExcludeList());
```
Comment (umamaheswararao) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestWritableECContainerProvider.java:274]: Same as above. Please few other places need to handle this.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestWritableECContainerProvider.java (hunk 1 above)
Comment (aswinshakil) [reply hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/pipeline/TestWritableECContainerProvider.java:239]: Thank for pointing it out. Updated the PR.

**Comment by aswinshakil** [comment]
@umamaheswararao Thanks for the review. I will update the null checks and other changes in a follow-up JIRA.

**Review by umamaheswararao — APPROVED** [review]
LGTM

**Comment by flyphant** [comment]
when i use cli command on the single node depolyment, input " bin/ozone sh key put /vol1/bucket1/key1 /tmp/myfile" , it's alse return the error: INTERNAL_ERROR Allocated 0 blocks. Requested 1 blocks
my version is 1.3.0, it's this bug fixed?

**Pull request merged**